README.md
# INTEL-SA-00075-Linux-Detection-And-Mitigation-Tools

Latest Release @ https://downloadcenter.intel.com/download/26799/INTEL-SA-00075-Linux-Detection-and-Mitigation-Tools

make

sudo ./INTEL-SA-00075-Discovery-Tool or sudo ./INTEL-SA-00075-Discovery-Tool -d /dev/mei

sudo ./INTEL-SA-00075-Unprovisioning-Tool or sudo ./INTEL-SA-00075-Unprovisioning-Tool -d /dev/mei


CHANGELOG
Release 1.0
INTEL-SA-00075-Discovery-Tool
INTEL-SA-00075-Unprovisioning-Tool
	- Added code documentation and centralized common functions for the tools.
====================================

Release 0.8
INTEL-SA-00075-Discovery-Tool
	- Tool should now support discovery on Intel(R) MEI firmware version 6.x
====================================

Release 0.5
INTEL-SA-00075-Discovery-Tool
INTEL-SA-00075-Unprovisioning-Tool
	- Tool not supported on systems with Intel(R) MEI firmware version 6.x
====================================



INTEL-SA-00075.h
/******************************************************************************
 * Intel-SA-00075.h
 *
 * This file is provided under a dual BSD/GPLv2 license.  When using or
 * redistributing this file, you may do so under either license.
 *
 * GPL LICENSE SUMMARY
 *
 * Copyright (C) 2003-2012, 2017 Intel Corporation. All rights reserved.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of version 2 of the GNU General Public License as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110,
 * USA
 *
 * The full GNU General Public License is included in this distribution
 * in the file called COPYING.
 *
 * Contact Information:
 *	Intel Corporation.
 *	linux-mei@linux.intel.com
 *	http://www.intel.com
 *
 * BSD LICENSE
 *
 * Copyright (C) 2003-2012, 2017 Intel Corporation. All rights reserved.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 *  * Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 *  * Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in
 *    the documentation and/or other materials provided with the
 *    distribution.
 *  * Neither the name Intel Corporation nor the names of its
 *    contributors may be used to endorse or promote products derived
 *    from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 *****************************************************************************/

#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <fcntl.h>
#include <sys/ioctl.h>
#include <unistd.h>
#include <errno.h>
#include <stdint.h>
#include <stdbool.h>
#include <bits/wordsize.h>
#include <linux/mei.h>

#define DEFAULT_MEI_DEV_NODE "/dev/mei0"

#define MEI_IAMTHIF UUID_LE(0x12f80028, 0xb4b7, 0x4b2d, 0xac, 0xa8, 0x46, 0xe0, \
 0xff, 0x65, 0x81, 0x4c)
#define MEI_MKHI_HIF  UUID_LE(0x8e6a6715, 0x9abc, 0x4043, 0x88, 0xef, 0x9e, 0x39,\
 0xc6, 0xf6, 0x3e, 0x0f)
#define MEI_MKHI_HIF_FIX UUID_LE(0x55213584, 0x9a29, 0x4916, 0xba, 0xdf, 0xf, \
0xb7, 0xed, 0x68, 0x2a, 0xeb)

/*****************************************************************************
 * Intel(R) MEI
 *****************************************************************************/
#define mei_err(_me, fmt, ARGS...) do {         \
	fprintf(stderr, "Error: " fmt, ##ARGS); \
} while (0)

struct mei {
    uuid_le guid;
    bool initialized;
    bool verbose;
    unsigned int buf_size;
    unsigned char prot_ver;
    int fd;
};

/***************************************************************************
 * Intel(R) AMT
 ***************************************************************************/

#define AMT_MAJOR_VERSION 1
#define AMT_MINOR_VERSION 1

#define AMT_STATUS_SUCCESS                  0x0
#define AMT_STATUS_INTERNAL_ERROR           0x1
#define AMT_STATUS_NOT_READY                0x2
#define AMT_STATUS_INVALID_AMT_MODE         0x3
#define AMT_STATUS_INVALID_MESSAGE_LENGTH   0x4

#define AMT_STATUS_HOST_IF_EMPTY_RESPONSE   0x4000
#define AMT_STATUS_SDK_RESOURCES            0x1004

typedef enum {
    AMT_CLIENT_TYPE = 1, MKHI_CLIENT_TYPE = 2, MKHI_FIX_CLIENT_TYPE = 3,
} CLIENT_TYPE;

struct heci_host_if {
    struct mei mei_cl;
    unsigned long send_timeout;
    bool initialized;
};

/*****************************************************************************
 * Check Provisioned State
 *****************************************************************************/
#define PROVISIONING_STATE_PRE 0
#define PROVISIONING_STATE_IN 1
#define PROVISIONING_STATE_POST 2

typedef struct {
    uint32_t Operation :23;
    uint32_t IsResponse :1;
    uint32_t Class :8;
} COMMAND_FMT;

typedef struct {
    uint8_t MajorNumber;
    uint8_t MinorNumber;
} PTHI_VERSION;

typedef struct {
    PTHI_VERSION Version;
    uint16_t Reserved;
    COMMAND_FMT Command;
    uint32_t Length;
} PTHI_MESSAGE_HEADER;

typedef struct {
    PTHI_MESSAGE_HEADER Header;
} CFG_GetProvisioningState_Request;

typedef struct {
    PTHI_MESSAGE_HEADER Header;
    uint32_t Status;
    uint32_t ProvisioningState;
} CFG_GetProvisioningState_Response;

/*****************************************************************************
 * MKHI Command constructs
 *****************************************************************************/
typedef struct {
    uint32_t GroupId :8;
    uint32_t Command :7;
    uint32_t IsResponse :1;
    uint32_t Reserved :8;
    uint32_t Result :8;
} MKHI_MESSAGE_HEADER;

typedef struct {
    MKHI_MESSAGE_HEADER Header;
} GEN_GET_VPRO_ALLOWED_Request;

typedef struct {
    MKHI_MESSAGE_HEADER Header;
    uint8_t VproAllowed;
} GEN_GET_VPRO_ALLOWED_Response;

/*****************************************************************************
 * Check Provisioning Control Mode
 *****************************************************************************/
typedef struct {
    PTHI_MESSAGE_HEADER Header;
} CFG_GetControlMode_Request;

typedef struct {
    PTHI_MESSAGE_HEADER Header;
    uint32_t Status;
    uint32_t ControlMode;  // returned upon success only
} CFG_GetControlMode_Response;

/*****************************************************************************
 * CCM UNPROVISION COMMAND
 *****************************************************************************/
typedef uint32_t AMT_STATUS;

typedef enum {
    CFG_PROVISIONING_MODE_NONE = 0,
    CFG_PROVISIONING_MODE_ENTERPRISE = 1,
    CFG_PROVISIONING_MODE_SMALL_BUSINESS = 2,
    CFG_PROVISIONING_MODE_REMOTE_CONNECTIVITY_SERVICE = 3,
} CFG_PROVISIONING_MODE;

typedef struct {
    PTHI_MESSAGE_HEADER Header;
    CFG_PROVISIONING_MODE Mode;
} CFG_Unprovision_Request;

typedef struct {
    PTHI_MESSAGE_HEADER Header;
    AMT_STATUS Status;
} CFG_Unprovision_Response;

/*****************************************************************************
 * Check UnProvisioning State
 *****************************************************************************/
typedef enum {
    CFG_UNPROVISIONING_STATE_NONE = 0, CFG_UNPROVISIONING_STATE_IN = 1,
} CFG_UNPROVISIONING_STATE;

typedef struct {
    PTHI_MESSAGE_HEADER Header;
} CFG_GetUnprovisioningState_Request;

typedef struct {
    PTHI_MESSAGE_HEADER Header;
    AMT_STATUS Status;
    CFG_UNPROVISIONING_STATE State;  // returned upon success only
} CFG_GetUnprovisioningState_Response;

/*****************************************************************************
 * Common Functions
 *****************************************************************************/

bool mei_init(struct mei *me, const uuid_le *guid, const char *device_path);

void mei_deinit(struct mei *cl);

ssize_t mei_recv_msg(struct mei *me, unsigned char *buffer, ssize_t len);

ssize_t mei_send_msg(struct mei *me, const unsigned char *buffer, ssize_t len);

int heci_send_recieve_message(const unsigned char *request, ssize_t req_len,
        unsigned char *response, ssize_t rsp_len, CLIENT_TYPE client_type,
        const char *device_path);

bool heci_host_if_init(struct heci_host_if *acmd, unsigned long send_timeout,
        const char *device_path, CLIENT_TYPE client_type);

int get_provisioning_control_mode(const char *device_path);

int get_provisioning_status(const char *device_path, bool unprovision);


INTEL-SA-00075-Unprovisioning-Tool.c
/******************************************************************************
 * Intel-SA-00075-Unprovisioning-Tool
 *
 * This file is provided under a dual BSD/GPLv2 license.  When using or
 * redistributing this file, you may do so under either license.
 *
 * GPL LICENSE SUMMARY
 *
 * Copyright (C) 2003-2012, 2017 Intel Corporation. All rights reserved.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of version 2 of the GNU General Public License as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110,
 * USA
 *
 * The full GNU General Public License is included in this distribution
 * in the file called COPYING.
 *
 * Contact Information:
 *	Intel Corporation.
 *	linux-mei@linux.intel.com
 *	http://www.intel.com
 *
 * BSD LICENSE
 *
 * Copyright (C) 2003-2012, 2017 Intel Corporation. All rights reserved.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 *  * Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 *  * Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in
 *    the documentation and/or other materials provided with the
 *    distribution.
 *  * Neither the name Intel Corporation nor the names of its
 *    contributors may be used to endorse or promote products derived
 *    from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 *****************************************************************************/

#include "INTEL-SA-00075.h"

/*****************************************************************************
 * Check UnProvisioning State
 *****************************************************************************/
/*
 * Function:     Reveals if the previous un-provisioning attempt was successful.
 *               Returns 1-->Un-Provisioning succeeded, 
 *                       0-->Un-Provisioning failed, 
 *                      -1-->Failures.
 * Dependencies: None
 * Arguments:    Intel(R) MEI kernel device node path
 * Description:  (1) Prepares command structure to send an GetUnprovisioningState 
 *                   request message to the connected Intel(R) AMT client in 
 *                   Intel(R) ME FW.
 *               (2) Calls the send receive message/ command function and  it 
 *                   records the unprovisioning state in the response structure 
 *                   memmber State.
 *                   OR--> Sets return rval to -1 if connection fails and jumps
 *                   to exit label failed_check_amt_unprovisioning_status
 *               (3) In either cases above, code now reaches the exit label where
 *                   it returns rval. 
 * Notes:        None
 */
int check_amt_unprovisioning_status(const char *device_path) {
    CFG_GetUnprovisioningState_Request request = { 
        .Header.Version.MajorNumber = 1, 
        .Header.Version.MinorNumber = 1, 
        .Header.Command.Class = 4,
        .Header.Command.Operation = 0x68, 
        .Header.Length = sizeof(request.Header) 
    };
    CFG_GetUnprovisioningState_Response response;

    CLIENT_TYPE client_type = AMT_CLIENT_TYPE;

    int rval = heci_send_recieve_message((const unsigned char *) &request,
            sizeof(request), (unsigned char *) &response, sizeof(response),
            client_type, device_path);

    if (rval < 0 || response.Status != AMT_STATUS_SUCCESS) {
        printf("Error: Failed Unprovisioning. Use Intel(R) MEBX to unprovision."
                " Or Contact OEM\n");
        rval = -1;
        goto failed_check_amt_unprovisioning_status;
    } else {
        if (!response.State) {
            printf("System needs to be unprovisioned.\n");
        } else {
            printf("System is already unprovisioned.\n");
            rval = -1;
        }
    }

failed_check_amt_unprovisioning_status: 
    return rval;
}

/*****************************************************************************
 * CCM UNPROVISION COMMAND
 *****************************************************************************/
/*
 * Function:     Attempts Full Intel (R) AMT UN-Provisioning in client mode/ CCM.
 * Returns:      0-->successful
 *              -1-->failure cases.
 * Dependencies: System is Intel(R) AMT provisioned in client mode.
 * Arguments:    
 * Description:  (1) Prepares command structure to send an CFG_Unprovisioning_Request 
 *                   request message to the connected Intel(R) AMT client in 
 *                   Intel(R) ME FW.
 *               (2) Calls the send receive message/ command function and  it 
 *                   records the unprovisioning response in the response structure 
 *                   memmber Status.
 *                   if Status is not AMT_STATUS_SUCCESS,
 *                   Sets return rval to -1 if connection fails and jumps
 *                   to exit label failed_un_provision_amt_ccm
 *               (3) In either cases above, code now reaches the exit label where
 *                   it returns rval.  
 * Notes:        None
 */
int un_provision_amt_ccm(const char *device_path) {
    CFG_PROVISIONING_MODE mode = CFG_PROVISIONING_MODE_NONE;
    CFG_Unprovision_Request request = { 
        .Header.Version.MajorNumber = AMT_MAJOR_VERSION, 
        .Header.Version.MinorNumber = AMT_MINOR_VERSION,
        .Header.Command.Class = 4, 
        .Header.Command.Operation = 0x10, 
        .Mode = mode, 
        .Header.Length = sizeof(request.Header) 
    };
    CFG_Unprovision_Response response;

    CLIENT_TYPE client_type = AMT_CLIENT_TYPE;

    int rval = heci_send_recieve_message((const unsigned char *) &request,
            sizeof(request), (unsigned char *) &response, sizeof(response),
            client_type, device_path);

    if (rval < 0 || response.Status != AMT_STATUS_SUCCESS) {
        printf(
                "Error: Failed Unprovisioning. Use Intel(R) MEBX to unprovision. "
                        "Or Contact OEM.");
        rval = -1;
        goto failed_un_provision_amt_ccm;
    } else {
        printf("\tSuccessfully Unprovisioned.\n");
    }

failed_un_provision_amt_ccm: 
    return rval;
}

/*****************************************************************************
 * INTEL-SA-00075-Unprovisioning-Tool Messages
 *****************************************************************************/
/*
 * Function:     Display tool banner along with copyright information.
 * Dependencies: None
 * Arguments:    None
 * Description:  None
 * Notes:        Ensure to update the release number every release.
 */
void print_tool_banner(void) {
    printf("\nINTEL-SA-00075-Unprovisioning-Tool -- Release 1.0\n");
    printf("Copyright (C) 2003-2012, 2017 Intel Corporation.  All rights reserved\n\n");
}

/*****************************************************************************
 * INTEL-SA-00075-Unprovisioning-Tool 
 *****************************************************************************/
/*
 * Function:     To un-provision client mode (CCM) AMT provisioned platform.
 * Dependencies: Root privilege/ permissions.
 * Arguments:    [INPUT] -d is the only expected command line option for 
 *               user defined /dev/mei# node
 * Description:  (1) Displays the tool version and copyright info
 *               (2) Ensures the tool was run with root permissions
 *               (3) Defaults to open /dev/mei0 node unless specified by user option
 *               (4) Checks if system is in provisioned state, if not - exits.
 *               (5) Checks if provisioning mode is ACM, if yes - exits.
 *                   Note it will attempt unprovisioning if mode is CCM/Unknown.
 *               (6) Checks if an UN-provisioning attempt has been made already.
 *                   Does nothing? (??Reboot-Message??) --Need this step?
 *               (7) Attempt UN-provisioning and exit if it fails.
 *               (8) Checks if the above UN-provisioning attempt go through.
 * Notes:        If everything went fine, run Discovery tool to ensure system is 
 *               unprovisioned/ not vulnerable. Note unprovisioning is a mitigation.
 */
int main(int argc, char **argv) {
    print_tool_banner();

    if (geteuid() != 0) {
        mei_err(me, "Please run the tool with root privilege.\n");
        exit(-1);
    }

    const char *dev_path;
    if (argc > 1 && !strncmp(argv[1], "-d", 2)) {
        dev_path = argv[2];
    } else {
        dev_path = DEFAULT_MEI_DEV_NODE;
    }

    printf("\n-----------------------------------------------------------\n");
    int ret = get_provisioning_status(dev_path, true);
    if (ret < 0) {
        goto out;
    }
    if (ret == 0) {
        printf("System is in unprovisioned state. Exiting.\n");
        goto out;
    }

    ret = get_provisioning_control_mode(dev_path);
    if (ret == 0) {
        printf("Control Mode: CLIENT / CCM\n");
    }
    if (ret == 2) {
        printf("Control Mode: ADMIN / ACM\n");
        printf("\tError: Cannot Unprovision - System provisioned in Admin Control"
                        " Mode.\n");
        printf("\tUnprovision via Intel(R) MEBX. Press CTRL+P during system boot."
                        " Or Contact OEM.\n");
        goto out;
    }
    if (ret == -1) {
        printf("Control Mode: Undetermined\n");
    }

    ret = check_amt_unprovisioning_status(dev_path);
    if (ret < 0) {
        goto out;
    }

    printf("Attempting Unprovisioning:\n");
    ret = un_provision_amt_ccm(dev_path);
    if (ret < 0) {
        goto out;
    }

    ret = check_amt_unprovisioning_status(dev_path);
    if (ret < 0) {
        goto out;
    }

out: 
    printf("\n----------------------------------------------------------\n");
    return ret;
}


INTEL-SA-00075.c
/******************************************************************************
 * Intel-SA-00075.c
 *
 * This file is provided under a dual BSD/GPLv2 license.  When using or
 * redistributing this file, you may do so under either license.
 *
 * GPL LICENSE SUMMARY
 *
 * Copyright (C) 2003-2012, 2017 Intel Corporation. All rights reserved.
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of version 2 of the GNU General Public License as
 * published by the Free Software Foundation.
 *
 * This program is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA 02110,
 * USA
 *
 * The full GNU General Public License is included in this distribution
 * in the file called COPYING.
 *
 * Contact Information:
 *	Intel Corporation.
 *	linux-mei@linux.intel.com
 *	http://www.intel.com
 *
 * BSD LICENSE
 *
 * Copyright (C) 2003-2012, 2017 Intel Corporation. All rights reserved.
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 *
 *  * Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 *  * Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in
 *    the documentation and/or other materials provided with the
 *    distribution.
 *  * Neither the name Intel Corporation nor the names of its
 *    contributors may be used to endorse or promote products derived
 *    from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 *****************************************************************************/

#include "INTEL-SA-00075.h"

/*****************************************************************************
 * Intel(R) MEI
 *****************************************************************************/

/*
 * Close the device handle to the Intel(R) MEI driver
 * And reset the initialized flag.
 */
void mei_deinit(struct mei *cl) {
    if (cl->fd != -1)
        close(cl->fd);
    cl->fd = -1;
    cl->buf_size = 0;
    cl->prot_ver = 0;
    cl->initialized = false;
}

/*
 * Attempts to open the dev node for Intel(R) MEI device driver and establish
 * connection with specified client uuid.
 * Returns a boolean: True if the handle was established and connection was made
 * successfully, False otherwise.
 */
bool mei_init(struct mei *me, const uuid_le *guid, const char *device_path) {
    int result;
    struct mei_client *cl;
    struct mei_connect_client_data data = { 0 };

    me->fd = open(device_path, O_RDWR);
    if (me->fd == -1) {
        mei_err(me, "%s %s\nCannot establish a handle to the Intel(R) MEI driver." 
            " Refer to Tool User Guide for more information.\n", strerror(errno), 
            device_path);
        exit(-1);
    }
    memcpy(&me->guid, guid, sizeof(*guid));
    me->initialized = true;

    memcpy(&data.in_client_uuid, &me->guid, sizeof(me->guid));
    result = ioctl(me->fd, IOCTL_MEI_CONNECT_CLIENT, &data);
    if (result) {
        mei_err(me, "IOCTL_MEI_CONNECT_CLIENT receive message. err=%d\n", result);
        goto err;
    }
    cl = &data.out_client_properties;

    me->buf_size = cl->max_msg_length;
    me->prot_ver = cl->protocol_version;

    return true;
err: 
    mei_deinit(me);
    return false;
}

/*
 * Writes out the command/ message to the initialized Intel(R) MEI client connection
 * with specified length of the message/ command.
 * If write goes through, returns number of bytes written otherwise the errno value.
 * After the message is sent it de-initializes the connection. 
 */
ssize_t mei_send_msg(struct mei *me, const unsigned char *buffer, ssize_t len) {
    ssize_t written;
    ssize_t rc;

    written = write(me->fd, buffer, len);
    if (written < 0) {
        rc = -errno;
        mei_err(me, "write failed with status %zd %s\n", written, strerror(errno));
        goto out;
    } else {
        rc = written;
    }
out: 
    if (rc < 0) {
        mei_deinit(me);
    }
    return rc;
}

/*
 * Retrieves response of specified length in the response buffer within a timeout
 * value specified, Any failures here would cause the code to exit since the Intel(R)
 * MEI clients are irresponsive. Otherwise return value is 0 for successful reads.
 */
ssize_t mei_recv_msg(struct mei *me, unsigned char *buffer, ssize_t len) {
    //receive message
    ssize_t rc = read(me->fd, buffer, len);

    if (rc < 0) {
        mei_err(me, "read operation failed with status %zd %s\n", rc, 
            strerror(errno));
        mei_deinit(me);
    }
    return rc;
}

/*****************************************************************************
 * HECI send/ receive message
 *****************************************************************************/
/*
 * Function:     Send a message to a specific HECI client and retrieve responses
 *               after establishing heci connection with the specified client.
 * Returns:      0 -->Successful sending message and receiving responses. 
 *              -1 -->Failures
 * Dependencies: Command structure is properly initialized
 * Arguments:    # Request/ Message - void* to accept different message structures.
 *               # Size of the request.
 *               # Response - void* to commonly record all types of responses.
 *               # Size of the response.
 *               # Intel(R) MEI HECI client to connect to send/ recevie message/responses
 *               # Intel(R) MEI kernel device node path
 * Description:  (1) Establishes connection with Intel (R) AMT client in Intel(R) MEI
 *                   firmware. If connection fails, Sets return rval to -1 and jumps
 *                   to exit label failed_check_amt_provisioning_status. 
 *               (2) Calls the send message/ command function and records the 
 *                   number of bytes written which should equal  size of request, 
 *                   if not it sets rval to -1 and jumps to exit label.
 *               (3) It records the number of bytes received in out_buf_sz variable. 
 *                   If out_buf_sz <= 0  it sets the rval to -1 and jumps to exit tag.
 *                   Otherwise it records the response structure. 
 *               (4) In either cases above, code now reaches the exit label, where
 *                   it disconnects the Intel(R) AMT client connection, closes the 
 *                   open dev/mei# node and returns rval. 
 * Notes:        None
 */
int heci_send_recieve_message(const unsigned char *request, ssize_t req_len,
        unsigned char *response, ssize_t rsp_len, CLIENT_TYPE client_type,
        const char *device_path) {
    //Send receive message to a heci client
    int rval = 0;
    struct heci_host_if acmd;
    if (!heci_host_if_init(&acmd, 5000, device_path, client_type)) {
        rval = -1;
        goto failed_heci_send_recieve_message;
    }

    uint32_t written = mei_send_msg(&acmd.mei_cl, request, req_len);
    if (written != req_len) {
        rval = -1;
        goto failed_heci_send_recieve_message;
    }

    uint32_t out_buf_sz = mei_recv_msg(&acmd.mei_cl, response, rsp_len);
    if (out_buf_sz <= 0) {
        rval = -1;
        goto failed_heci_send_recieve_message;
    }
failed_heci_send_recieve_message: 
    mei_deinit(&acmd.mei_cl);
    return rval;
}

/***************************************************************************
 * Intel(R) AMT -  Host Interface
 ***************************************************************************/
/*
 * Passes the information required to connect to specific Intel(R) MEI clients.
 * The function is commonly called by the tools prior to sending commands to the
 * specified clients and also sets up the timeout values.
 */
bool heci_host_if_init(struct heci_host_if *acmd, unsigned long send_timeout,
        const char *device_path, CLIENT_TYPE client_type) {
    acmd->send_timeout = (send_timeout) ? send_timeout : 20000;
    switch (client_type) {
    //Connect appropriate client
    case AMT_CLIENT_TYPE:
        acmd->initialized = mei_init(&acmd->mei_cl, &MEI_IAMTHIF, device_path);
        break;
    case MKHI_CLIENT_TYPE:
        acmd->initialized = mei_init(&acmd->mei_cl, &MEI_MKHI_HIF, device_path);
        break;
    case MKHI_FIX_CLIENT_TYPE:
        acmd->initialized = mei_init(&acmd->mei_cl, &MEI_MKHI_HIF_FIX,
                device_path);
        break;
    default:
        printf("DEBUG %d\n", client_type);
        acmd->initialized = false;
        break;
    }
    return acmd->initialized;
}

/*****************************************************************************
 * Check Provisioning Control Mode
 *****************************************************************************/
/*
 * Function:     Retrieves the mode in which the Intel(R) AMT was provisioned.
 *               Returns 1-->CCM, 
 *                       2-->ACM, 
 *                       0-->Undetermined, 
 *                      -1-->Failures.
 * Dependencies: None
 * Arguments:    Intel(R) MEI kernel device node path
 * Description:  (1) Prepares command structure to send an GetControlMode request
 *                   message to the connected Intel(R) AMT client in Intel(R) ME FW.
 *               (2) Calls the send receive message/ command function 
 *                   and records the provisioning control mode in the 
 *                   response structure member ControlMode.
 *               (3) In either cases above, code now reaches the exit label, where
 *                   it returns rval. 
 * Notes:        None
 */
int get_provisioning_control_mode(const char *device_path) {
    CFG_GetControlMode_Request request = { 
        .Header.Version.MajorNumber = 1,
        .Header.Version.MinorNumber = 1, 
        .Header.Command.Class = 0x4,
        .Header.Command.Operation = 0x6B, 
        .Header.Length = sizeof(request.Header) 
    };
    CFG_GetControlMode_Response response;

    CLIENT_TYPE client_type = AMT_CLIENT_TYPE;
    int rval = heci_send_recieve_message((const unsigned char *) &request,
            sizeof(request), (unsigned char *) &response, sizeof(response),
            client_type, device_path);

    if (rval < 0 || response.Status != AMT_STATUS_SUCCESS) {
        rval = -1;
        goto failed_get_provisioning_control_mode;
    } else {
        rval = response.ControlMode;
    }

failed_get_provisioning_control_mode: 
    return rval;
}

/*****************************************************************************
 * Check Provisioned State
 *****************************************************************************/
/*
 * Function:     Reveals if the previous un-provisioning attempt was successful.
 *               Returns 0-->PROVISIONING_STATE_PRE 
 *                       1-->PROVISIONING_STATE_IN 
 *                       2-->PROVISIONING_STATE_POST
 *                      -1-->Failures
 * Dependencies: None
 * Arguments:    # Intel(R) MEI kernel device node path
 *               # Flag to track tool message handling for the discovery and unprovisioning
 * Description:  (1) Prepares command structure to send an GetProvisioningState request
 *                   message to the connected Intel(R) AMT client in Intel(R) ME FW.
 *               (2) Calls the send receive message/ command function 
 *                   and records the provisioning status in the  response structure 
 *                   member ProvisioningState.
 *               (3) In either cases above, code now reaches the exit label, where
 *                   it returns rval.
 * Notes:        None
 */
int get_provisioning_status(const char *device_path, bool unprovision) {
    CFG_GetProvisioningState_Request request = {
        .Header.Version.MajorNumber = 1, 
        .Header.Version.MinorNumber = 1,
        .Header.Command.Class = 4, 
        .Header.Command.Operation = 0x11,
        .Header.Length = sizeof(request.Header) 
    };
    CFG_GetProvisioningState_Response response;

    CLIENT_TYPE client_type = AMT_CLIENT_TYPE;
    int rval = heci_send_recieve_message((const unsigned char *) &request,
            sizeof(request), (unsigned char *) &response, sizeof(response),
            client_type, device_path);

    if (rval < 0 || response.Status != AMT_STATUS_SUCCESS) {
        rval = -1;
        if (unprovision) {
            printf("Error: Failed Unprovisioning. Use Intel(R) MEBX to "
                "unprovision. Or Contact OEM\n");
        } else {
            printf("Error: Failed to retrieve response for provisioning status:"
                    " %08X\n", response.Status);
        }
        goto failed_check_amt_provision_status;
    } else {
        if (response.ProvisioningState == PROVISIONING_STATE_PRE) {
            printf("PROVISIONING_STATE = PRE\n"); //Not Provisioned
            rval = 0;
        }
        if (response.ProvisioningState == PROVISIONING_STATE_IN) {
            printf("PROVISIONING_STATE = IN\n");
            rval = 1;
        }
        if (response.ProvisioningState == PROVISIONING_STATE_POST) {
            printf("PROVISIONING_STATE = POST\n"); //Provisioned
            rval = 2;
        }
    }
failed_check_amt_provision_status: 
    return rval;
}
